Implement speculative prerendering

Swanand01 opened the pull request:
## Summary

This PR makes use of the [Speculation Rules API](https://developer.mozilla.org/en-US/docs/Web/API/Speculation_Rules_API) to enable dynamically prerendering editor and story URLs.

## User-facing changes

<!--
Please describe your changes.
Include before/after screenshots or a short video.
-->
Prerendering for these URLs has been enabled for the following pages:
1. Dashboard - Create New Story, Edit story
2. All Stories - Create New Story, Edit story, View story

## Testing Instructions

This PR can be tested by following these steps:

1. Go to the Stories Dashboard page.
2. Open the Application Tab and go to Speculative loads > Speculations.
3. A list of URLs with their Status should appear.
4. Hover over a story to see the status change to Ready.

## WIP

1. Add prerendering on the front end for story permalink when hovering over the link to view the story
2. Add a check to see if the Speculation Rules plugin is active


## Reviews

### Does this PR have a security-related impact?
No.

### Does this PR change what data or activity we track or use?
🤔

### Does this PR have a legal-related impact?
No.

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it
- [ ] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [ ] I have added a matching `Type: XYZ` label to the PR

---

Fixes #13603

Swanand01 committed:
Add the `Speculative_Prerendering` service class

Swanand01 committed:
Update `SERVICES` to list the `Speculative_Prerendering` service

Swanand01 committed:
Fix: Update 2020 to 2024 and fix phpstan linting issue

swissspidy commented on includes/Speculative_Prerendering.php:
```diff
@@ -0,0 +1,114 @@
+
+declare(strict_types = 1);
+
+namespace Google\Web_Stories\Admin;
```
```suggestion
namespace Google\Web_Stories;
```

As per my other comment. Let's remove the `Admin` namespace here as this will be used on the frontend as well.

swissspidy commented on includes/Plugin.php:
```diff
@@ -140,6 +140,7 @@ class Plugin extends ServiceBasedPlugin {
 		'user_preferences'             => User\Preferences::class,
 		'remove_transients'            => Remove_Transients::class,
 		'web_stories_block'            => Block\Web_Stories_Block::class,
+		'speculative_prerendering'     => Admin\Speculative_Prerendering::class,
```
```suggestion
		'speculative_prerendering'     => Speculative_Prerendering::class,
```

Let's remove the `Admin` namespace here as this will be used on the frontend as well.

swissspidy commented on includes/Speculative_Prerendering.php:
```diff
@@ -0,0 +1,114 @@
+	private array $matches = [
+		'dashboard'   => [ 'post.php?post=*&action=edit', 'post-new.php?post_type=web-story' ],
+		'all_stories' => [ 'post.php?post=*&action=edit', 'post-new.php?post_type=web-story', '/web-stories*' ],
+		'archive'     => [ '/web-stories*' ],
```
If the site URL and home URL are different, a relative URL like `/web-stories*` won't necessarily work.

Also, the `web-stories/` permalink can be customized by the user to something different, or even be disabled. So we should not hardcode it. That's why we usually use `get_post_type_archive_link( $this->story_post_type->get_slug() )` to get that path. The function returns `false` if the archive page is disabled.

We need to find a solution for this eventually.

Swanand01 replied:
Maybe we could default to `/web-stories/*` if `get_post_type_archive_link()` returns `false`. I had the archive page disabled, and yet the stories were visible on `/web-stories/*`

swissspidy replied:
Ah right, of course, it's for the single views, not the archive. Then let's use at least `Story_Post_Type::REWRITE_SLUG` here instead of hardcoding `web-stories`

swissspidy commented on includes/Speculative_Prerendering.php:
```diff
@@ -0,0 +1,114 @@
+	 */
+	public function load_rules( string $hook ): void {
+		$rules = [];
+		if ( 'web-story_page_stories-dashboard' === $hook ) {
```
Instead of hardcoding the hook suffix like this, we should inject the `Dashboard` service and get it from `$this->dashboard->get_hook_suffix( 'stories-dashboard' )`

Swanand01 replied:
`$this->dashboard->get_hook_suffix( 'stories-dashboard' )` is always returning `false` for me. 

```php
public function get_hook_suffix( string $key ) {
  return $this->hook_suffix[ $key ] ?? false;
}
```
Upon taking a closer look, `$this->hook_suffix[ $key ]` gets us `web-story_page_stories-dashboard`, but the function returns `false`

Swanand01 replied:
The doc string has - 

```
@return bool|string The dashboard page's hook_suffix, or false if the user does not have the capability required.
```

Maybe this has something to do with user capabilities.

swissspidy replied:
You'll need to add `Dashboard::class` to `\Google\Web_Stories\Plugin::get_shared_instances()` in order to address this. Otherwise the `Dashboard` instance you get is a new one which at this point won't have registered the menu item yet. Think of this as turning `Dashboard` into a singleton.

swissspidy commented on includes/Speculative_Prerendering.php:
```diff
@@ -0,0 +1,114 @@
+		}
+
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended -- Nonce verification not needed as we're just checking for the 'web-story' post type.
+		if ( 'edit.php' === $hook && isset( $_GET['post_type'] ) && 'web-story' === $_GET['post_type'] ) {
```
We should be able to use `$this->context->is_story_editor()` and `'edit' === $this->context->get_screen_base()` to detect whether we're on the "All Stories" page.

See `\Google\Web_Stories\Admin\Dashboard::display_link_to_dashboard` for precedent.

swissspidy reviewed (commented):
Thanks for sharing a first draft! This already seems to work quite well and makes for near-instant navigations to single stories or the story editor. 👏 🚀

Swanand01 committed:
Remove the `Admin` namespace

Swanand01 committed:
Use Context and Dashboard services to dynamically build URLs

swissspidy commented on includes/Plugin.php:
```diff
@@ -140,6 +140,7 @@ class Plugin extends ServiceBasedPlugin {
 		'user_preferences'             => User\Preferences::class,
 		'remove_transients'            => Remove_Transients::class,
 		'web_stories_block'            => Block\Web_Stories_Block::class,
+		'speculative_prerendering'     => Speculative_Prerendering::class,
```
New services also need to be added in `.phpstorm.meta.php`

Swanand01 committed:
Add `Dashboard` class to `Plugin::get_shared_instances()`

Swanand01 committed:
Use `story_post_type::REWRITE_SLUG` for view story URLs

Swanand01 committed:
Add `Speculative_Prerendering` service class to `.phpstorm.meta.php`

Swanand01 committed:
Use `elseif` instead of `if`

Swanand01 committed:
Add test for `load_rules`

Swanand01 committed:
Fix: Update priority to 10

Swanand01 committed:
Move instance creation and mock object creation to `set_up`

Swanand01 committed:
Remove types from class variables and make call

Swanand01 committed:
Make `print_rules` and `get_rules` private

Swanand01 committed:
Check if `print_rules` was called

Swanand01 committed:
Revert: Make print_rules and get_rules private

Also remove URLs for archive page

Swanand01 committed:
Add tests for `get_rules`

swissspidy commented on tests/phpunit/integration/tests/Speculative_Prerendering.php:
```diff
@@ -0,0 +1,176 @@
+		parent::set_up();
+
+		$this->dashboard       = $this->createMock( Dashboard::class );
+		$this->story_post_type = $this->createMock( Story_Post_Type::class );
```
Not necessary to use a mock for this class here. Can just use `$this->injector->make( Story_Post_Type::class )`

Swanand01 replied:
Can the same be done for the context and dashboard?

Swanand01 replied:
Also, please expect some more commits on this PR 😅. Anurag and I spent a while trying to get a working local test environment, but it didn't happen. So, we can only run the tests on the PR.

swissspidy replied:
Mhm maybe.

If you do it for `Dashboard`, then you'll just have to call `$this->dashboard->add_menu_page()` to the hook suffix exists.

If you do it for `Context`, then you'll have do something like `$GLOBALS['current_screen'] = convert_to_screen( Story_Post_Type::POST_TYPE_SLUG );` to set the current screen, so that the context methods return the correct thing.

Swanand01 replied:
In that case, I suppose we could do it just for `$story_post_type` for now.

Swanand01 replied:
Using `$this->injector->make( Story_Post_Type::class )` for `$story_post_type` throws an error here:

https://github.com/GoogleForCreators/web-stories-wp/blob/289baa482e974ff9d870c0b3d88b8bd7580d1733/tests/phpunit/integration/tests/Speculative_Prerendering.php#L96

```
Undefined method 'method'
```
Possibly because this `method` exists on the `MockObject`

swissspidy replied:
Well, `$this->injector->make( Story_Post_Type::class )` creates a proper instance of the class, not a mock.
Without a mock class, you also don't need to mock the `get_slug` method. You can just remove this line.

Swanand01 committed:
Use `injector` to build `story_post_type`

Swanand01 committed:
Merge branch 'main' into feat/13603-speculative-prerendering

AnuragVasanwala committed:
Merge branch 'main' into feat/13603-speculative-prerendering

AnuragVasanwala committed:
Merge branch 'main' into feat/13603-speculative-prerendering

swissspidy commented on includes/Speculative_Prerendering.php:
```diff
@@ -0,0 +1,168 @@
+/**
+ * Speculative_Prerendering class.
+ */
+class Speculative_Prerendering extends Service_Base {
```
Looking at this again, let's rename this everywhere to `Speculation_Rules`

Swanand01 committed:
Rename Speculative_Prerendering to Speculation_Rules

Swanand01 committed:
Update Speculation_Rules.php

- Remove the Dashboard injection
- Hook print_rules into admin_footer and wp_footer.
- Remove generate_matches, get_matches_for_page and load_rules
- In get_rules() add an apply_filters() call

Swanand01 committed:
Add speculation rules in Dashboard using the new filter

Swanand01 committed:
Update test for `get_rules`

Swanand01 committed:
Add test for `print_rules`

swissspidy commented on .phpstorm.meta.php:
```diff
@@ -85,6 +85,7 @@
 			'remove_transients'            => \Google\Web_Stories\Remove_Transients::class,
 			'web_stories_block'            => \Google\Web_Stories\Block\Web_Stories_Block::class,
 			'injector'                     => \Google\Web_Stories\Infrastructure\Injector::class,
+			'speculation_rules'     => \Google\Web_Stories\Speculation_Rules::class,
```
Nit: incorrect indentation

swissspidy commented on includes/Admin/Dashboard.php:
```diff
@@ -573,4 +574,32 @@ public function display_link_to_dashboard(): void {
+					'where'     => [
+						'and' => [
+							[
+								'href_matches' => [ $edit_story_url, $new_story_url ],
```
We can add `$view_story_url` here too so that links to the story permalink can be preloaded.

We should also test  whether `$view_story_url` should be in `prefetch` instead, not `prerender`, given that it's usually loaded in a lightbox anyway.

Swanand01 replied:
Tested both `prefetch` and `prerender` for `$view_story_url`, and I couldn't see a noticeable difference. Prerender appeared to be just slightly faster, in terms of page loading times.

swissspidy commented on includes/Admin/Dashboard.php:
```diff
@@ -573,4 +574,32 @@ public function display_link_to_dashboard(): void {
+
+	/**
+	 * Load the Speculation Rules for the Dashboard.
+	 *
```
We usually add `@since` annotations to all methods, at least the public ones. The next release will be `1.37.0`.

Same goes for the `Speculation_Rules` plugin as well.

swissspidy commented on includes/Speculation_Rules.php:
```diff
@@ -0,0 +1,111 @@
+				],
+			];
+		}
+		return apply_filters( 'web_stories_speculation_rules', $rules );
```
We usually add docblocks for all our filters, including `@since` tags. Please check other hooks in the plugin for how it's done.

swissspidy commented on includes/Speculation_Rules.php:
```diff
@@ -0,0 +1,111 @@
+	 *
+	 * @return array An array containing prerendering rules.
+	 */
+	public function get_rules(): array {
```
This can be private

swissspidy commented on tests/phpunit/integration/tests/Speculation_Rules.php:
```diff
@@ -0,0 +1,125 @@
+
+		$actual_rules = $prerendering_class->get_rules();
+		$this->assertEquals( $expected_rules, $actual_rules );
+	}
```
This test is a bit pointless as it mocks `get_rules` to return X and then you verify whether it returns X.

We already have `$this->instance` from `set_up` above, so let's use it.

```suggestion
	/**
	 * @covers ::print_rules
	 */
	public function test_print_rules(): void {
		$expected_rules = [
			'prerender' => [
				[
					'source'    => 'document',
					'where'     => [
						'and' => [
							[ 'href_matches' => [ Story_Post_Type::REWRITE_SLUG, sprintf( '/%s/*', $this->story_post_type::REWRITE_SLUG ) ] ],
						],
					],
					'eagerness' => 'moderate',
				],
			],
		];

		$output = get_echo( [ $this->instance, 'print_rules' ] );

		$actual_rules = $this->instance->get_rules();
		$this->assertStringContainsString( wp_json_encode( $expected_rules ), $output );
	}
```

swissspidy commented on includes/Admin/Dashboard.php:
```diff
@@ -573,4 +574,32 @@ public function display_link_to_dashboard(): void {
+	 *
+	 * @return array An array containing prerendering rules.
+	 */
+	public function load_speculation_rules(): array {
```
"Load" is maybe the wrong term. Perhaps `filter_` or `add_`?

swissspidy replied:
Turns out this code is still running on the frontend. Apparently the `Dashboard` class is still instantiated on the frontend too.

To fix this:

1. Make the class implement the `Conditional` interface
2. Add the following class method

```php
	/**
	 * Check whether the conditional object is currently needed.
	 *
	 * @since 1.37.0
	 *
	 * @return bool Whether the conditional object is needed.
	 */
	public static function is_needed(): bool {
		return is_admin();
	}
```

This way the `Dashboard` class is truly only instantiated in the admin.

Not a fault of this PR, this was an existing issue.

swissspidy commented on includes/Speculation_Rules.php:
```diff
@@ -0,0 +1,111 @@
+			return;
+		}
+
+		$encoded_rules = wp_json_encode( $rules );
```
I think we need to use the `JSON_UNESCAPED_SLASHES` flag here

swissspidy commented on includes/Admin/Dashboard.php:
```diff
@@ -573,4 +574,32 @@ public function display_link_to_dashboard(): void {
+						'and' => [
+							[
+								'href_matches' => [ $edit_story_url, $new_story_url ],
+							],
```
I think `href_matches` must not be an array, but a string. Can you test that?

Might need to do this instead:

```
"and": [
        { "href_matches": "..." },
        { "href_matches": "...." },
        ....
]
```

Swanand01 replied:
The rules don't appear in the Application panel if we do 

```
"and": [
        { "href_matches": "..." },
        { "href_matches": "...." },
        ....
]
```

But they show up with our existing configuration.
I suppose `href_matches` can take an array.

Swanand01 committed:
Fix indentation

Swanand01 committed:
Add the since tag

Swanand01 committed:
Remove test for `get_rules` and modify test for `print_rules`

Swanand01 committed:
Rename `load_speculation_rules` to `add_speculation_rules`

Swanand01 committed:
Use `JSON_UNESCAPED_SLASHES` in `wp_json_encode`

Swanand01 committed:
Make the `Dashboard` class implement `Conditional`

Swanand01 committed:
Add what the array consists of in doc comment

swissspidy commented on tests/phpunit/integration/tests/Speculation_Rules.php:
```diff
@@ -0,0 +1,79 @@
+		$this->instance->get_rules();
+
+		$encoded_rules = wp_json_encode( $expected_rules );
+		if ( false !== $encoded_rules ) {
```
Let's just cast to a string so we don't have to do this `false` check here.

swissspidy replied:
Alternatively, instead of the if/else, can do `$this->assertNotFalse( $encoded_rules )`

swissspidy commented on tests/phpunit/integration/tests/Speculation_Rules.php:
```diff
@@ -0,0 +1,79 @@
+		$output         = get_echo( [ $this->instance, 'print_rules' ] );
+		$this->instance->get_rules();
+
+		$encoded_rules = wp_json_encode( $expected_rules );
```
Probably need `JSON_UNESCAPED_SLASHES` here as well

Swanand01 committed:
Fix: Use `assertNotFalse` in test

Swanand01 committed:
Merge branch 'main' into feat/13603-speculative-prerendering

Swanand01 committed:
Simplify speculation rules, and add `$view_story_url` to Dashboard rule

Swanand01 committed:
SImplify rules in test

Swanand01 committed:
Skip flaky test

Swanand01 committed:
Add `eslint-disable-next-line` to the skipped test

Swanand01 committed:
Merge branch 'main' into feat/13603-speculative-prerendering

Swanand01 committed:
Use fixture.events.sleep() in the failing test

Swanand01 committed:
Merge branch 'feat/13603-speculative-prerendering' of https://github.com/GoogleForCreators/web-stories-wp into feat/13603-speculative-prerendering

AnuragVasanwala committed:
Merge branch 'main' into feat/13603-speculative-prerendering

swissspidy committed:
Merge branch 'main' into feat/13603-speculative-prerendering

Swanand01 committed:
Merge branch 'main' into feat/13603-speculative-prerendering

Swanand01 committed:
Merge branch 'main' into feat/13603-speculative-prerendering

Swanand01 committed:
Merge branch 'main' into feat/13603-speculative-prerendering

swissspidy committed:
Merge branch 'main' into feat/13603-speculative-prerendering
